### PR TITLE
Experimental variant of BINARY_ADD specialization.

### PR DESCRIPTION
**Pull request by markshannon**

### CODE REVIEW

File: Python/ceval.c [hunk 1]
```diff
@@ -2182,6 +2191,30 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
+            PyObject *left = SECOND();
+            PyObject *right = TOP();
+            DEOPT_IF(Py_TYPE(left)->tp_version_tag != cache0->left_version, BINARY_ADD);
+            DEOPT_IF(Py_TYPE(right)->tp_version_tag != cache0->right_version, BINARY_ADD);
+            record_cache_hit(cache0);
+            binaryfunc function = caches[-1].func_ptr.function;
+            PyObject *sum = function(left, right);
+            if (sum == Py_NotImplemented) {
+                Py_DECREF(sum);
+                sum = NULL;
```
Comment (Fidget-Spinner) [inline Python/ceval.c:2185]: You may want to consider replacing this with
```suggestion
                Py_SETREF(sum, NULL);
```

**Review by Fidget-Spinner — COMMENTED** [review]
I wonder if we'll see any speedups from this. I don't see how it will be much faster than ``PyNumber_Add``. If the compiler's smart about it, it'll probably inline ``PyNumber_Add`` during PGO, and we'll have almost the same function

File: Python/ceval.c (hunk 1 above)
Comment (brandtbucher) [reply Python/ceval.c:2185]: Or:

```suggestion
                Py_CLEAR(sum);
```

**Comment by markshannon** [comment]
Will need reworking once #29482 is merged.

**Comment by markshannon** [comment]
@brandtbucher Maybe something like this will show better results now that the binary and inplace operators have been merged?

**Pull request closed without merging**